### PR TITLE
Differentiability

### PR DESCRIPTION
**Pull request by wbenoit26**
@chreissel I think this PR should address the issues with differentiating waveforms. Can you take a look and confirm? I also fixed an in-place modification in `spectral.py`, so I _think_ the whole chain should be differentiable now.

### Benchmarks (A30)

**Waveforms**

```
| Benchmark                         |  Baseline |   Current | Delta          |
|-----------------------------------|-----------|-----------|----------------|
| test_taylorf2_forward[batch_32]   |   5.11 ms |   5.16 ms | Stable (+1.1%) |
| test_taylorf2_forward[batch_128]  |   5.12 ms |   5.16 ms | Stable (+0.8%) |
| test_taylorf2_forward[batch_512]  |   9.09 ms |   9.08 ms | Stable (-0.1%) |
| test_phenomd_forward[batch_32]    |  39.25 ms |  39.51 ms | Stable (+0.7%) |
| test_phenomd_forward[batch_128]   |  43.96 ms |  44.09 ms | Stable (+0.3%) |
| test_phenomd_forward[batch_512]   | 116.61 ms | 116.50 ms | Stable (-0.1%) |
| test_phenompv2_forward[batch_32]  |  45.37 ms |  45.69 ms | Stable (+0.7%) |
| test_phenompv2_forward[batch_128] |  47.86 ms |  48.05 ms | Stable (+0.4%) |
| test_phenompv2_forward[batch_512] | 115.71 ms | 115.61 ms | Stable (-0.1%) |
```

**normalize_by_psd**

```
| Benchmark                        |  Baseline |   Current | Delta             |
|----------------------------------|-----------|-----------|-------------------|
| test_normalize_by_psd[batch_32]  | 185.91 μs | 164.74 μs | Improved (-11.4%) |
| test_normalize_by_psd[batch_128] | 490.82 μs | 433.22 μs | Improved (-11.7%) |
| test_normalize_by_psd[batch_512] |   1.97 ms |   1.70 ms | Improved (-13.7%) |
```

### CODE REVIEW

**Review by deepchatterjeeligo — COMMENTED** [review]
I'll let Christina comment whether her use case is met, but I'm satisfied. I also checked that not just scalar gradient but also full partial derivatives work out fine. For example, ${\partial h \over \partial M_c}  (f)$ i.e. the partial along chirp mass as a function of frequency:
```
>>> def h(chirp_mass, freqs):
...     params = {"chirp_mass": chirp_mass,
...         "mass_ratio": torch.tensor([0.5], dtype=FLOAT64, requires_grad=True),
...         "s1x": torch.tensor([0.5], dtype=FLOAT64, requires_grad=True),
...         "s1y": torch.tensor([0.5], dtype=FLOAT64, requires_grad=True),
...         "s1z": torch.tensor([0.5], dtype=FLOAT64, requires_grad=True),
...         "s2x": torch.tensor([0.5], dtype=FLOAT64, requires_grad=True),
...         "s2y": torch.tensor([0.5], dtype=FLOAT64, requires_grad=True),
...         "s2z": torch.tensor([0.5], dtype=FLOAT64, requires_grad=True),
...         "distance": torch.tensor([100.0], dtype=FLOAT64, requires_grad=True),
...         "phic": torch.tensor([0.0], dtype=FLOAT64, requires_grad=True),
...         "inclination": torch.tensor([0.5], dtype=FLOAT64, requires_grad=True),
...     }
...     hc_ml4gw, hp_ml4gw = waveforms.IMRPhenomPv2()(freqs,f_ref=f_ref,**params)
...     return hc_ml4gw.real.squeeze()
>>> torch.autograd.functional.jacobian(lambda mc: h(mc, torch_freqs), m_chirp)
tensor([[-1.0557e-21],
        [ 1.4928e-21],
        [-1.5487e-21],
        [-1.4766e-21],
        [-1.2393e-21],
.
.
.
>>> torch_freqs.shape
torch.Size([80])
>>> torch.autograd.functional.jacobian(lambda mc: h(mc, torch_freqs), m_chirp).shape
torch.Size([80, 1])
```
@wbenoit26 can we add one example (maybe similar to the above) of jacobian calculation to the unittests as well.

**Comment by wbenoit26** [comment]
Just realized - the `TimeDomainCBCWaveformGenerator` also needs a few changes as well.

**Comment by wbenoit26** [comment]
I was wrong, the TD generator is fine. But I added a test to confirm it.

**Review by deepchatterjeeligo — APPROVED** [review]
LGTM

**Comment by chreissel** [comment]
thanks for adding this feature! I can confirm that I am able to use ml4gw now without any further modifications and can differentiate in all parameter directions.

**Comment by wbenoit26** [comment]
Great! I'll merge this in and make a release.
